common/numerated/src/iterators.rs

```diff
@@ -0,0 +1,330 @@
+                    debug_assert!(self.interval1.is_some(), "`T: Numerated` impl error");
+                } else if interval1.end() == interval2.end() {
+                    self.interval1 = None;
+                    self.interval2 = None;
+                } else {
+                    self.interval1 = None;
+                }
+
+                if let Some(new_end) = interval2.start().dec_if_gt(interval1.start()) {
+                    let res = Interval::new(interval1.start(), new_end);
+                    debug_assert!(res.is_some(), "`T: Numerated` impl error");
+                    return res;
+                } else {
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+/// Helper struct to iterate over voids in tree.
+///
+/// See also [`IntervalsTree::voids`].
+pub struct VoidsIterator<T: Numerated, I: Iterator<Item = Interval<T>>> {
+    pub(crate) inner: Option<(I, Interval<T>)>,
+}
+
+impl<T: Numerated, I: Iterator<Item = Interval<T>>> Iterator for VoidsIterator<T, I> {
+    type Item = Interval<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (iter, interval) = self.inner.as_mut()?;
+        if let Some(next) = iter.next() {
+            let (start, end) = next.into_parts();
+
+            // Guaranties by tree: between two intervals always exists void.
+            debug_assert!(interval.start() < start);
+
+            let Some(void_end) = start.dec_if_gt(interval.start()) else {
+                debug_assert!(false, "`T: Numerated` impl error");
+                return None;
+            };
+
+            let res = Interval::new(interval.start(), void_end);
+            debug_assert!(res.is_some(), "`T: Numerated` impl error");
+
+            if let Some(new_start) = end.inc_if_lt(interval.end()) {
+                let Some(new_interval) = Interval::new(new_start, interval.end()) else {
+                    debug_assert!(false, "`T: Numerated` impl error");
+                    return None;
+                };
+                *interval = new_interval;
+            } else {
+                self.inner = None;
+            }
+
+            res
+        } else {
+            let res = Some(*interval);
+            self.inner = None;
+            res
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use alloc::vec;
+
+    use super::*;
+
+    #[test]
+    fn empty_interval() {
+        assert!(IntervalIterator::<u8>::new_empty().is_empty());
+        assert!(IntervalIterator::<u8>::new_with_len(1, 0)
+            .unwrap()
+            .is_empty());
+        assert!(IntervalIterator::<u8>::new_with_len(None, 0)
+            .unwrap()
+            .is_empty());
+        assert!(IntervalIterator::<u8>::try_from(1..1).unwrap().is_empty());
+        assert!(IntervalIterator::<u8>::try_from(..0).unwrap().is_empty());
+        assert!(IntervalIterator::<u8>::try_from(None..).unwrap().is_empty());
+    }
+
+    #[test]
+    fn voids_iterator() {
```

Comment:
this task could be solved by difference iterator, isn't it?, so voids iterator is kinda useless

Reply:
Difference iterator, according to docs is "Helper struct to iterate over intervals from `tree1`, which are not in `tree2`.", in fact this should be "iterator of intervals" instead of tree1/2
BUT
voids iterator is difference of [..] interval and a given one, isn't it?